src/foam/core/Action.js

```diff
@@ -134,59 +134,58 @@ foam.CLASS({
   ],
 
   methods: [
-    function andSlots(a, b) {
-      return this.ExpressionSlot.create({
-        args: [ a, b ],
-        code: function(a, b) {
-          return a && b;
-        }
-      });
-    },
 
-    function andSlotAndPromise(slot, promise) {
-      return this.andSlots(slot, this.PromiseSlot.create({
-        promise: promise
-      }));
+    function hasPermissions_(x, property) {
+      if ( ! this.permissionRequired || ! x.auth ) return Promise.resolve(true);
+      return Promise.all(this.permissionConfig[property].map(p => {
+        return x.auth.check(null, p);
+      })).then(arr => arr.every(b => b));
     },
 
-    function checkPermission(x, permission) {
-      if ( ! this.permissionConfig || ! x.auth ) return Promise.resolve(true);
-      return x.auth.check(null, permission);
-    },
-
-    function isEnabledFor(data) {
-      return this.isEnabled
-        ? foam.Function.withArgs(this.isEnabled, data)
-        : true;
-    },
-
-    function createIsEnabled$(data$) {
-      return this.createSlot_(this.isEnabled, 'enabled', data$);
+    function createPermissionlessCheckFor_(data$, property) {
+      if ( ! this[property] ) return foam.core.ConstantSlot.create({value: true});
+      return foam.core.ExpressionSlot.create({
+        obj$: data$,
+        code: this[property]
+      });
     },
 
-    function isAvailableFor(data) {
-      return this.isAvailable
-        ? foam.Function.withArgs(this.isAvailable, data)
-        : true;
+    function createIsEnabled$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
```

Comment:
Can we avoid creating the PromiseSlot in the case where we _don't_ have permissions or an auth service?  If we leave it as a ConstantSlot then we can save even more listeners from being hooked up in what is currently a common case of no permissions on an action.